### PR TITLE
sqlproxyccl: change denylist into an access control list

### PR DESCRIPTION
**Pull request by pjtatlow**
To support an IP Allowlist in the sqlproxy, this change extends
the denylist code to make the Watcher support multiple AccessControllers.
Each AccessController is consulted before allowing a connection through,
and rechecked on any changes to the underlying files.

Part of: [CC-8136](https://cockroachlabs.atlassian.net/browse/CC-8136)

### CODE REVIEW

File: pkg/ccl/sqlproxyccl/acl/watcher.go [hunk 1]
```diff
@@ -173,14 +160,32 @@ func (l *listener) Less(than btree.Item) bool {
...
-func checkConnection(connection ConnectionTags, list *Denylist) error {
-	ip := DenyEntity{Item: connection.IP, Type: IPAddrType}
-	if err := list.Denied(ip); err != nil {
-		return errors.Wrapf(err, "connection ip '%v' denied", connection.IP)
-	}
-	cluster := DenyEntity{Item: connection.Cluster, Type: ClusterType}
-	if err := list.Denied(cluster); err != nil {
-		return errors.Wrapf(err, "connection cluster '%v' denied", connection.Cluster)
+func checkListeners(
+	listeners *btree.BTree, timesource timeutil.TimeSource, controllers ...AccessController,
```
[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/acl/watcher.go:164]: Ah my thought here was that by passing them in individually rather than the slice, I would avoid races. But you're right, I would need to make a copy of the slice anyways so I will just do that instead.

File: pkg/ccl/sqlproxyccl/denylist/file_test.go [hunk 2]
```diff
@@ -1,253 +0,0 @@
-// Copyright 2021 The Cockroach Authors.
```
[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/denylist/file_test.go:1]: TIL 🤯

File: pkg/ccl/sqlproxyccl/acl/file.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	pollingInterval time.Duration
+	timeSource      timeutil.TimeSource
+}
+
+// newAccessControllerFromFile returns a AccessController that automatically watches updates to a file.
+// Note: this currently does not return an error. This is by design, since even if we trouble
+// initiating an access controller with file, we can always update the file with correct content during
+// runtime. We don't want sqlproxy fail to start just because there's something wrong with
+// contents of an access control file.
+// TODO(pj): Does the above still apply to the AllowList? Or was that specific to the DenyList?
```
[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/acl/file.go:44]: Do you think failing to load the denylist should have the same behavior as failing to load the allowlist?

File: pkg/ccl/sqlproxyccl/acl/watcher.go [hunk 4]
```diff
@@ -173,14 +160,32 @@ func (l *listener) Less(than btree.Item) bool {
...
+		}
+		return true
+	})
+}
+
+func checkConnection(
+	connection ConnectionTags, timesource timeutil.TimeSource, controllers ...AccessController,
+) error {
+	for _, c := range controllers {
+		if c != nil {
```
[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/acl/watcher.go:184]: In the case where there is an error reading the files on startup, it will return a `nil` AccessController. But if we're going to fail to start in that case, we can remove this check.

File: pkg/ccl/sqlproxyccl/denylist/file_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/denylist/file_test.go:1]: Had to make two commits to do this, but hopefully this looks better now. Github is still showing those files as deleted, but my local git blame seems to be working correctly.

File: pkg/ccl/sqlproxyccl/acl/file.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/acl/file.go]: Spoke with Jeff, decided we want the sqlproxy to fail on startup if it has any invalid configurations (allow or deny list).

File: pkg/ccl/sqlproxyccl/denylist/file_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaylim-crl) [reply pkg/ccl/sqlproxyccl/denylist/file_test.go:1]: By the way, IIRC, in CRDB land, if we move files like that, we should also make sure that the individual commits are compilable to aid git bisect. i.e. `package denylist` should be renamed to `package acl` in the move commit. Someone else could probably confirm.

[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/denylist/file_test.go:1]: Oof okay, that's going to take a lot more changes then. Let me try again.

File: pkg/ccl/sqlproxyccl/acl/watcher.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+			return nil, err
+		}
+		w.addAccessController(c, next)
+	}
+	return w, nil
+}
+
+func (w *Watcher) addAccessController(controller AccessController, next chan AccessController) {
+	index := len(w.controllers)
+	w.controllers = append(w.controllers, controller)
```
Comment (JeffSwenson) [inline pkg/ccl/sqlproxyccl/acl/watcher.go:143]: There is still a small race condition here. It's unlikely to be hit since the channel won't emit a new access controller for five seconds, but we should handle it. addAccessController does not lock controllers and updateAccessController copies the list.

Normally it is safe to assume the lock in the function constructing an object, but that assumption is broken if any of the methods called in the constructor spawns goroutines.

**Review by jeffswenson — COMMENTED** [review]
This is looking great! Once the race is fixed I'll approve the change.

File: pkg/ccl/sqlproxyccl/acl/access_control.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2021 The Cockroach Authors.
```
[the remark this answers is not in the document]
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/acl/access_control.go]: Good call!

File: pkg/ccl/sqlproxyccl/acl/watcher.go (hunk 5 above)
Comment (pjtatlow) [reply pkg/ccl/sqlproxyccl/acl/watcher.go:146]: It took me a second to see it, but you're totally right. Fixed!

**Review by jeffswenson — APPROVED** [review]
LGTM
